library/ssl_cli.c

```diff
@@ -2711,7 +2711,7 @@ static int ssl_parse_certificate_request( mbedtls_ssl_context *ssl )
      * therefore the buffer length at this point must be greater than that
      * regardless of the actual code path.
      */
-    if( ssl->in_hslen <= mbedtls_ssl_hs_hdr_len( ssl ) + 2 + n )
+    if( ssl->in_hslen <= mbedtls_ssl_hs_hdr_len( ssl ) + 3 + n )
```

Comment:
@catenacyber @mpg @sbutcher-arm What is the justification for this? To me it seems that the previous check was correct.

Reply:
Indeed, maybe this was about the problem with 
```
            MBEDTLS_SSL_DEBUG_MSG( 3, ( "Supported Signature Algorithm found: %d"
                                        ",%d", sig_alg[i], sig_alg[i + 1]  ) );
```
fixed by https://github.com/catenacyber/mbedtls/commit/bc231cc9b0cecf9723e9e12a6de2cc78f847e1b6

Reply:
@hanno-arm  It seems to me that if `ssl->in_hslen == mbedtls_ssl_hs_hdr_len( ssl ) + 2 + n` then access to `buf[mbedtls_ssl_hs_hdr_len( ssl ) + 2 + n]` on line 2727 below isn't properly guarded.

Reply:
@mpg My point is that the code fails even if `ssl->in_hslen == mbedtls_ssl_hs_hdr_len( ssl ) + 3 + n` which seems unnecessary.

Reply:
If that's the case and we're below TLS 1.2, and `dn_len == 0`, then we might fail to parse a valid `CertificateRequest` message. This case is explicitly allowed in the [TLS 1.1 RFC 4346, Section-7.4.4](https://tools.ietf.org/html/rfc4346#section-7.4.4):

```
      certificate_authorities
         A list of the distinguished names of acceptable certificate
         authorities.  These distinguished names may specify a desired
         distinguished name for a root CA or for a subordinate CA; thus,
         this message can be used to describe both known roots and a
         desired authorization space.  If the certificate_authorities
         list is empty then the client MAY send any certificate of the
         appropriate ClientCertificateType, unless there is some
         external arrangement to the contrary.
```

Reply:
Ok, so that was the "not enough tea" branch: I missed that the test if `<=` not `<`. Indeed please disregard my previous comment, I agree with your analysis.